src/OrbitGl/CaptureViewElement.cpp

```diff
@@ -127,6 +127,10 @@ void CaptureViewElement::SetVisible(bool value) {
   RequestUpdate();
 }
 
+float CaptureViewElement::HorizontalClamp(float pos_x) const {
+  return std::clamp(pos_x, pos_[0], pos_[0] + width_);
```

Comment:
Interesting. I assume that you are doing this so events are not drawn outside of the area of their track.

This is not-consistent with what is happening in the right_margin (and actually won't work there because the size of the box can overflow the margin). Also a line outside of the left-most part might be visible misleading the user. In the right-margin area we are rendering an entire rectangle with background_color on top, so events in that area won't be visible since they are behind the rectangle. So:

1- Either instead of clamping, you do the same than the right-margin and draw another rectangle for the `left_margin`.

2- Aim for a proper solution that will work for both margins to later erase the right_margin.
I have some ideas about this "proper solution" but I'm also interested in @reichlfl opinion.

I thought on doing this on the "Batcher/PrimitiveAssembler" level. Instead of the current implementation, all methods to add primitives will take an additional Pos/Size parameter from the container that will limit the area where the primitives will be drawn (and possibly no draw anything at all). It's a big refactor though but it will be error-prone as we won't need to clamp everything before drawing.

What do you think? @reichlfl @pierricgimmig

Reply:
2 would be my preferred solution as it neatly scales to all elements, and you don't need to take care about it when implementing draw methods.

But I'm not convinced that adding it as parameters to PrimitiveAssembler calls is the right thing to do, it's *a lot* of parameters that we'd need to add all across the place, its a big refactor and takes a lot of additional mental overhead when writing code. Instead, I would add this into the Batcher state, akin to `Push/PopTranslation`. Each `CaptureViewElement` pushes its bounding rect onto the batcher before drawing, and pops it afterwards, that way the draw methods are basically unaffected. It may still be quite a bit of work to verify everything still works as expected, and it might also have significant performance impact.

So in conclusion, I think this localized fix is also fine for me. But I would change a couple of things that I missed in my first review, sorry for that (thanks for pointing me to this, Pablo):

- For consistency, you should clamp both the position and the width - probably the latter needs a separate method as it needs to take pos + width into account
- I would give these methods a more descriptive name
- I would add them to `Track` instead of `CaptureViewElement` to make it clear that this is - for now - a change that is somewhat locally to the tracks.

Reply:
> Each CaptureViewElement pushes its bounding rect onto the batcher before drawing, and pops it afterwards, that way the draw methods are basically unaffected.

Right! I forgot about Translations. Your idea was to push the size in addition to the position as it's currently working?

> and it might also have significant performance impact.

Why?

> So in conclusion, I think this localized fix is also fine for me. 

I'm not sure if I agree of having a temporal solution just for Tracks. I feel that it will add some mess to the code that might be harder to understand in the future. My first idea of just adding a margin rectangle is more consistent with the current mess while we plan this Translation proper solution.

Reply:
I implemented Florian's suggestions. @karupayun, please see rationale below about removing margin rectangles.